routing/DELEGATED_NAMING_HTTP.md

```diff
@@ -0,0 +1,176 @@
+# Delegated Naming HTTP API
+
+![wip](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square)
+
+**Author(s)**:
+
+- [Masih H. Derkani](https://github.com/masih)
+
+**Maintainer(s)**:
+
+* * *
+
+**Abstract**
+
+Delegated naming is a mechanism for IPFS implementations to offload naming system to another process
+or server. This includes naming resolution as well as publication of new naming records. This
+document describes an HTTP API through which such functionality is facilitated.
+
+## API Specification
+
+The Delegated Naming System HTTP API uses the `application/json` content type by default.
+
+As such, human-readable encodings of types are preferred. This spec may be updated in the future
+with a compact `application/cbor` encoding, in which case compact encodings of the various types
+would be used.
+
+## Common Data Types
+
+### IPNS Record
+
+The following snippet outlines the JSON schema of IPNS records:
+
+```json
+{
+  "Signature": "<signature>",
+  "Payload": {
+    "Value": "<value>",
+    "Sequence": 0,
+    "Validity": {
+      "EOL": {
+        "Timestamp": 0,
+        "AdvisoryTTL": 0
+      }
+    },
+    "PublicKey": "<optional-public-key>",
+    "ExtendedData": {}
+  }
+}
+```
+
+Where:
+
+- `Signature` is the multibase-encoded signature of the sha256 hash of the `Payload` field, signed
+  using the private key that corresponds to the `PublicKey` in the `Payload` if present. And
+  Otherwise, the private key associcated to the IPNS record key. Signing details for specific key
+  types should
+  follow [libp2p/peerid specs](https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#key-types),
+  unless stated otherwise.
+- `Payload` is the content of the IPNS record as specified
+  by [IPNS Record](../ipns/IPNS.md#ipns-record) specification:
```

Comment:
I see this as an opportunity to reduce cognitive cycles for users that are trying to make sense if IPNS records. I am not sure I understand the necessity to carry over identical binary formatting onto specifications that are written to _delegate_ services.

> introduce computation or storage cost (either you store CBOR and generate JSON, or you store both)
> Do we need JSON for IPNS?

Dlegated IPNI servers will most likely be unpacking the records before storage because the chances are they won't be using the vanilla DHT for storage. Aside from that, the specification proposed here offers backwards compatibility with the old binary format, while allowing the plain text human readable format as well. 

Mixing both `COBR` and `JSON` in a request response mechanism that clearly supports `Content-Type` mechanism to communicate formatting seems less than optimum. As I elaborated in other comments, it seems extreme to carry on with the existing binary format regardless when native HTTP apis can support both with added benefits of better UX, human readability and cross language/platform integration (e.g. inserting IPNS records from the browser over webtransport as an example).

We could have used the binary format used in DHT for all of the delegated routing HTTP APIs. But we did not for good reason: human readability, ease of use across multiple platforms/environments, easier debugging and so on. It seems like an opportunity missed to enforce the old formats in HTTP APIs where native support for extensible formats are built in. 

Sure we can make support for old formats mandatory instead of optional but it is unclear why that should be the _only_ format supported?

Reply:
I fully agree we want to minimize cognitive overhead, and  you are right, mixing JSON and CBOR  in `application/json` response will be against that. I am concerned about having deserialized, UNVERIFIED value right there in JSON payload, but if IPNI is okay with that, I am fine with it as well, as long clients can opt-in to verifiable binary format.

My _reductio ad absurdum_ failed to get the point about signatures across, so I'll try to rephrase:
All I want, is to make sure we have understanding that **we can't invent a new signature type for IPNS**.
So far, this IPIP  states that `Signature` in JSON response type invents new signature type:

> `Signature` [in JSON response from IPNI] is the multibase-encoded signature of the sha256 hash of the Payload field, signed using the private key that corresponds to the PublicKey in the Payload if present. And Otherwise, the private key associcated to the IPNS record key. Signing details for specific key types should follow libp2p/peerid specs, unless stated otherwise.

The way `signatureV2` is generated is defined in [IPNS spec](https://github.com/ipfs/specs/blob/main/ipns/IPNS.md#record-creation) and all clients expect this. Inventing anything new here means you force clients to now support two competing IPNS specs, and do CPU work twice by creating two different signatures, and to know when use which. 

This is harmful to the ecosystem, where we already struggle to clean up and [finish transition from V1 to V2](https://github.com/ipfs/specs/issues/376) by removing requirement for V1 fields. We want to have a single signature, and a simple format. 

The only path forward is to update description of `Signature` field in JSON representation and state it should match `IpnsEntry.signatureV2` from [IPNS spec](https://github.com/ipfs/specs/blob/main/ipns/IPNS.md#record-creation).

Reply:
Resolving this, as we've decided to reduce the scope of this initial IPIP and only support [application/vnd.ipfs.ipns-record](https://www.iana.org/assignments/media-types/application/vnd.ipfs.ipns-record).